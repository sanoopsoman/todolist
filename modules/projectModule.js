import { v4 as uuidv4 } from 'uuid'

/**
 * The projectModule is a module that provides functions for managing projects and todos.
 *
 * @namespace projectModule
 */
const projectModule = (() => {
    const STORAGE_KEY = 'projectApp.projects';

    let projects = JSON.parse(localStorage.getItem(STORAGE_KEY) || '[]');

    const saveToLocalStorage = () => {
        localStorage.setItem(STORAGE_KEY, JSON.stringify(projects));
    }

    const addProject = (title) => {
        const project = {
            id: uuidv4(),
            title: title,
            todos: []
        }

        projects.push(project);
        saveToLocalStorage();

        return project;
    }

    const updateProject = (id, title) => {
        const projectIndex = projects.findIndex(project => project.id === id);
        if (projectIndex !== -1) {
            projects[projectIndex].title = title;
            saveToLocalStorage();
        } else {
            throw new Error("Project not found.");
        }
        return projects[projectIndex];
    }

    const getProjects = () => projects;

    const addTodoToProject = (projectId, todoId) => {
        const projectIndex = projects.findIndex(project => project.id === projectId);
        if (projectIndex !== -1) {
            projects[projectIndex].todos.push(todoId);
            saveToLocalStorage();
        } else {
            throw new Error('Project not found.');
        }
    }


    const getTodos = (projectId) => {
        const project = projects.find(project => project.id === projectId);
        return project ? project.todos : [];
    }

    const deleteProjects = (projectId) => {
        const index = projects.findIndex(project => project.id === projectId);

        if (index !== -1) {
            projects.splice(index, 1);
            saveToLocalStorage();
        } else {
            throw new Error("Project not found.");
        }
    }

    return {
        addProject,
        updateProject,
        getProjects,
        getTodos,
        addTodoToProject,
        deleteProjects
    }
 })()

export default projectModule;