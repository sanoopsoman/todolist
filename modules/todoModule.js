import { v4 as uuidv4 } from 'uuid';
/**
 * todoModules - A module for managing todo items.
 *
 * @module todoModules
 */
const todoModules = (() => {
    const STORAGE_KEY = 'todoApp.todo';

    let todoList = JSON.parse(localStorage.getItem(STORAGE_KEY) || '[]');

    /**
     * Saves the todoList to the local storage using the provided key.
     *
     * @function saveToLocalStorage
     * @returns {void}
     */
    const saveToLocalStorage = () => {
        localStorage.setItem(STORAGE_KEY, JSON.stringify(todoList));
    };

    const addTodo = (title, description, dueDate, priority, projectId) => {
        const todo = {
            id: uuidv4(),
            title,
            description,
            dueDate,
            priority,
            projectId,
            status: 'todo'
        };
        todoList.push(todo);
        saveToLocalStorage();

        return todo;
    };

    const getTodos = (id) => {
        return todoList.filter(todo => todo.projectId === id);
    };

    const getTodo = (id) => {
        return todoList.find(todo => todo.id === id)
    }

    const updateTodoStatus = (id, newStatus) => {
        const index = todoList.findIndex(todo => todo.id === id);
        if (index !== -1) {
            todoList[index].status = newStatus;
            saveToLocalStorage();
        } else {
            throw new Error ('Todo not found');
        }
    }

    const deleteTodo = (id) => {
        const index = todoList.findIndex(todo => todo.id === id);
        if (index !== -1) {
            todoList.splice(index, 1);
            saveToLocalStorage();
        }
    };

    return {
        addTodo,
        getTodos,
        getTodo,
        updateTodoStatus,
        deleteTodo,
    };
})();

export default todoModules;