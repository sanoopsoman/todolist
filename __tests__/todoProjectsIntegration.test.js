import { describe, it, expect, beforeEach } from 'vitest';
import todoModule from '../modules/todoModule';
import projectModule from '../modules/projectModule';

describe('Integration between todoModule and projectModule', () => {
    beforeEach(() => {
        // Clear all entries before each test
        localStorage.clear();
        // Reset modules if necessary, depending on how they are implemented
    });

    it('should allow adding todos to a project and retrieve them accurately', () => {
        // Step 1: Create a new project
        const newProject = projectModule.addProject('Work Project');
        //const projects = projectModule.getProjects();
        //const projectId = projects[0].id;

        // Step 2: Add todos to the project
        const newTodo = todoModule.addTodo('Finish report', 'Complete the annual report.', '2023-04-15', 'high', 'Include financials', ['Draft', 'Review', 'Finalize'], newProject.id);
        // const todo = todoModule.getTodos()[0];

        // Step 3: Add the todo to the project
        projectModule.addTodoToProject(newProject.id, newTodo.id);

        // Step 4: Check that the todo has been added to the project
        const projectTodos = projectModule.getTodos(newProject.id);
        expect(projectTodos).toHaveLength(1);
        expect(projectTodos[0]).toBe(newTodo.id);

        // Fetch the todo object using the ID from todoModule and check its details
        const todoDetails = todoModule.getTodo(projectTodos[0]); // Assuming this method exists
        expect(todoDetails.title).toBe('Finish report'); // Check the title of the todo

        // Verify full integration
        const projects = projectModule.getProjects();
        expect(projects.find(p => p.id === newProject.id).todos).toContain(newTodo.id);
    });
});
