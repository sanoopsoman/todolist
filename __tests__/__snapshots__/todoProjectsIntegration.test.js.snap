// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration between todoModule and projectModule > should allow adding todos to a project and retrieve them accurately 1`] = `
[
  "6f9bdda3-da73-41d8-9d80-b5946ad88ca5",
]
`;
